src/core/toPath.js

```diff
@@ -102,139 +110,84 @@ const allowingArrays = fn => arg => {
-        if (str.charAt(index) === '.') index++
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
-
-          } else { // If a slice quote is found
-
-            // Fetch slice start and end, and parse them as slice indexes (empty or valid int)
-            const sliceStart = arrayIndexValue.substring(0, sliceDelimIndex), sliceEnd = arrayIndexValue.substring(sliceDelimIndex + 1)
-            const nSliceStart = toSliceIndex(sliceStart), nSliceEnd = toSliceIndex(sliceEnd)
-
-            // Add array index to path, as a slice if both slice indexes are valid (undefined or int), or as a string
-            path.push(isSliceIndex(nSliceStart) && isSliceIndex(nSliceEnd) ? [nSliceStart, nSliceEnd] : arrayIndexValue)
-          }
-        }
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-        // Stop if end of string has been reached
-        if (index === str.length) break
-      }
-    }
+/**
+ * @typedef {function(string): string[]} Matcher a function that can replace String.prototype.match
+ * @param {string} str string to match against
+ * @param {[(Matcher | RegExp), function(string[]): *]} matchers
+ *   pairs of a regexp to match str against, and a function to transform the resulting match object into the final result
+ * @param {*} defaultResult
+ *   value to return if no matcher matches
+ * @returns {*} output value of the first cond that matches or defaultResult if no cond matches
+ */
+const match = (str, matchers, defaultResult) => {
+  for (const [matcher, mapper] of matchers) {
+    const match = matcher instanceof RegExp ? str.match(matcher) : matcher(str)
+    if (match) return mapper(match.slice(1))
+  }
+  return defaultResult
+}
 
+match.andCheck = (matcher, predicate) => {
+  return str => {
+    const match = str.match(sliceNotation)
+    if (!match) return match
+    return predicate(match.slice(1)) ? match : null
   }
+}
 
-  return path
+const quotedBracketNotation = /^\[(['"])(.*?[^\\])\1\]?\.?(.*)$/
+const incompleteQuotedBracketNotation = /^\[["'](.*)$/
+const sliceNotation = /^\[([^:\]]*):([^:\]]*)\]\.?(.*)$/
+const bareBracketNotation = /^\[([^\]]*)\]\.?(.*)$/
+const incompleteBareBracketNotation = /^\[(.*)$/
+const pathSegmentEndedByDot = /^([^.[]*?)\.(.*)$/
+const pathSegmentEndedByBracket = /^([^.[]*?)(\[.*)$/
+
+const stringToPath2 = str => {
+  return match(str, [
```

Comment:
I would extract mappers in functions with a meaningful name. What do you think about it ?

Reply:
What would be meaningful names in this case? Let's take slice notation as an example: `processSliceNotation`? `sliceNotationToPath`? `parseSliceNotation`?

Reply:
parseSliceNotation is fine !

Reply:
Following https://github.com/Zenika/immutadot/pull/115#discussion_r150529285, I can either do

```js
const matchSomething = ... // regexp or function
const parseSomething = ... // function

match([
  makeMatcher(matchSomething, parseSomething),
  ...,
])
```

or

```js
const handleSomething = makeMatcher(/* inline matcher and parser */)

match([
  handleSomething,
  ...,
])
```

What do you prefer?

Reply:
I prefer the second one :+1:

handleSomething is fine. With a concrete example:

```
const matchSegmentEndedByBracket = ... // regex
const parseSegmentEndedByBracket = ... // function
const handleSegmentEndedByBracket = makeMatcher(matchSegmentEndedByBracket, parseSegmentEndedByBracket)

match([
  handleSegmentEndedByBracket,
  ...,
])
```